lib/database/player-achievement.php

```diff
@@ -0,0 +1,610 @@
+        $retVal['Error'] = "Unofficial achievements cannot be unlocked";
+        return $retVal;
+    }
+
+    $hasAwardTypes = playerHasUnlock($user, $achIDToAward);
+    $hasRegular = $hasAwardTypes['HasRegular'];
+    $hasHardcore = $hasAwardTypes['HasHardcore'];
+    $alreadyAwarded = $isHardcore ? $hasHardcore : $hasRegular;
+
+    $awardedOK = true;
+    if ($isHardcore && !$hasHardcore) {
+        $awardedOK &= insertAchievementUnlockIntoAwardedTable($user, $achIDToAward, true);
+    }
+    if (!$hasRegular && $awardedOK) {
+        $awardedOK &= insertAchievementUnlockIntoAwardedTable($user, $achIDToAward, false);
+    }
+
+    if (!$awardedOK) {
+        $retVal['Error'] = "Could not unlock achievement for player";
+        return $retVal;
+    }
+
+    if (!$alreadyAwarded) {
+        // testFullyCompletedGame could post a mastery notification. make sure to post
+        // the achievement unlock notification first
+        postActivity($user, ActivityType::EarnedAchievement, $achIDToAward, $isHardcore);
+    }
+
+    $completion = testFullyCompletedGame($achData['GameID'], $user, $isHardcore, !$alreadyAwarded);
+    if (array_key_exists('NumAwarded', $completion)) {
+        $retVal['AchievementsRemaining'] = $completion['NumAch'] - $completion['NumAwarded'];
+    }
+
+    if ($alreadyAwarded) {
+        // XXX: do not change the messages here. the client detects them and does not report
+        // them as errors.
+        if ($isHardcore) {
+            $retVal['Error'] = "Player already unlocked this achievement in hardcore mode";
+        } else {
+            $retVal['Error'] = "Player already unlocked this achievement";
+        }
+
+        return $retVal;
+    }
+
+    $pointsToGive = $achData['Points'];
+    settype($pointsToGive, 'integer');
+
+    if ($isHardcore && !$hasRegular) {
+        // Double points (award base as well!)
+        $pointsToGive *= 2;
+    }
+
+    $query = "UPDATE UserAccounts SET RAPoints=RAPoints+$pointsToGive, Updated=NOW() WHERE User='$user'";
+    $dbResult = s_mysql_query($query);
+    if (!$dbResult) {
+        $retVal['Error'] = "Could not add points for this player";
+        return $retVal;
+    }
+
+    $retVal['Success'] = true;
+    // Achievements all awarded. Now housekeeping (no error handling?)
+
+    static_setlastearnedachievement($achIDToAward, $user, $achData['Points']);
+
+    if ($user != $achData['Author']) {
+        attributeDevelopmentAuthor($achData['Author'], $pointsToGive);
+    }
+
+    // Update GameData
+    // Removed: this needs rethinking! //##SD TBD
+    // recalculateTrueRatio( $gameID );    // Heavy!
+    // Add TA to the player for this achievement, NOW that the TA value has been recalculated
+    // Select the NEW TA from this achievement, as it has just been recalculated
+    $query = "SELECT TrueRatio
+              FROM Achievements
+              WHERE ID='$achIDToAward'";
+    $dbResult = s_mysql_query($query);
```

Comment:
`TrueRatio` is returned as part of `GetAchievementMetadataJSON`. This query is just to get the updated value, which didn't occur on line 137.

Reply:
Hrm, true, that has to be a left-over. I guess it would be ok to remove all of the TruePoints related code as it couldn't have had any impact for quite some time?

Reply:
Not all. Just up through line 150. The next query will update the white points for the user, just using the current TrueRatio value.

Reply:
Makes sense, changed it accordingly